### ISSUE TITLE
Probles with training

### ISSUE THREAD

**Issue by Palkos83** [MatheusMRFM/A3C-LSTM-with-Tensorflow#3]
Hi 

Firstly, really nice code. It helped me to understand A3C fundamentals. However I do struggle to get it to converge. I tried at least four different implementations and most of them are having this same issue. 
Yesterday I tried your branch and kicked off the pong training out of the box. The only change I made was to use 8 workers as my rig has 8 cores. Sadly even after 1.8M of epochs it is not converging. The agent barely moves the towards the ball. 

Here is a screenshot from Tensorboard:
![image](https://user-images.githubusercontent.com/20494744/32981405-580922f4-cc6d-11e7-9c22-e798ee02e390.png)

I came across your notes on stackoverflow where you seemed to have sorted the issue by adding small number to the policy output avoiding NaNs. 
https://stackoverflow.com/questions/44926583/cant-get-my-a3c-with-lstm-layer-using-tensorflow-to-work
However that do not seem to match your current code.

Any ideas what might be the problem? Or how to even start debugging it to try to determine that?
Many thanks for help in advance.

**Comment by MatheusMRFM** [comment]
Hello there! First of all, thanks for using my code! I'm glad it helped you better understand the A3C method.

Regarding your problem, I think the issue here is that you left it running only for a small amount of time. Before I uploaded this code, I remember using it to train an agent for Pong (both Pong-v4 and PongDeterministic-v4) and Breakout-v4. Both achieved good results. For instance, here is the reward curve for the PongDeterministic: 

![7e-4 30 84](https://user-images.githubusercontent.com/18707264/32981604-cd7fe162-cc5e-11e7-8b0e-73d3f441dc1a.png)

and here is for Pong-v4:

![7e-4 30 84](https://user-images.githubusercontent.com/18707264/32981611-e1ae7e1e-cc5e-11e7-9b24-fda7c2c1461d.png)

Notice that the agent only starts to increase its score somewhere around 3M frames. Therefore, try leaving the agent training for longer and see what happens.

But just to make sure, I will try running my github version to see if there is any problem. I return to when I  have my new results.

As for my thread in stackoverflow, I ended up using the "tf.clip_by_value" to avoid zeros in the policy, which in turn avoids NAN when passing the policy to the log function (line 196 of Network.py).

**Comment by Palkos83** [comment]
Many thanks, for super fast answer. I will try leaving it for longer than. Just out of curiosity, how long does it take you to train 1M steps? I was running the code on I7 7700K + GTX1080 GPU (i am not sure if your code benefits from it though) and 1.8M steps took nearly a day.  That was with 8 workers running as well.

Is that expected?

Once again, many thanks for help.

**Comment by MatheusMRFM** [comment]
I left the code running right after I read your first question (which was about 3 hours ago) and, at the moment, I'm already at 2.695.268 steps (score still remains approximately -21). Therefore, I think that perhaps there is some problem with your tensorflow installation. My guess is the number of workers you are using: I made a quick search and it seems that [your CPU has in fact 4 processors, not 8](https://ark.intel.com/products/97129/Intel-Core-i7-7700K-Processor-8M-Cache-up-to-4_50-GHz). In this case, I recommend using only 4 workers, in order to use 1 thread per processor (although your processor seems to have hyper-threading). My CPU is also an i7 quad core (don't remember the exact name), and using 4 workers usually results in faster runs.

Make this experiment and see if works. As a final comment, note that my code doesn't support GPU, because I have no CUDA device in my PC :(

**Comment by Palkos83** [comment]
Many thanks for response. You are right it is 4 physical cores and 8 threads. I reduced the number of workers back to 4 and turned off the rendering of the workers as well (this is what made the most of the difference I presume as it started to go through the steps much faster - about 2M steps per hour). 

And, the best news is - it trained as you predicted, all I had to be patient :)

![image](https://user-images.githubusercontent.com/20494744/32989822-97c13ce4-cd15-11e7-8bf1-ee539cc45f71.png)

That is great, working really nice. I will try breakout and other games now as well.

Great job, nice code, very helpful!

**Comment by MatheusMRFM** [comment]
Glad to hear that this had a happy ending :)

I forgot to ask if you left the 'render' flag to True. Indeed this uses a lot of extra processing when compared to running without rendering.

Feel free to ask if you have any other questions.
